packages/flutter/lib/src/widgets/async.dart

```diff
@@ -0,0 +1,256 @@
+
+  FutureBuilder({Key key, this.future, this.builder}) : super(key: key);
+
+  @override
+  State<FutureBuilder<T>> createState() => new _FutureBuilderState<T>();
+}
+
+class _FutureBuilderState<T> extends State<FutureBuilder<T>> {
+  Stream<T> _stream;
+
+  @override
+  void initState() {
+    super.initState();
+    _subscribe();
+  }
+
+  @override
+  void didUpdateConfig(FutureBuilder<T> oldConfig) {
+    if (!identical(oldConfig.future, config.future)) {
+      _unsubscribe();
+      _subscribe();
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) =>
+      new StreamBuilder<T>(stream: _stream, builder: config.builder);
+
+  @override
+  void dispose() {
+    _unsubscribe();
+    super.dispose();
+  }
+
+  void _subscribe() {
+    _stream = config.future?.asStream();
+  }
+
+  void _unsubscribe() {
+    _stream = null;
+  }
+}
+
+/// Widget that builds itself based on the latest snapshot of interaction with
+/// the specified [stream].
+///
+/// The building strategy is defined by the given [builder].
+class StreamBuilder<T> extends StreamFold<T, AsyncSnapshot<T>> {
+  final AsyncWidgetBuilder<T> builder;
+
+  StreamBuilder({Key key, Stream<T> stream, this.builder})
+      : super(key: key, stream: stream);
+
+  @override
+  AsyncSnapshot<T> initial() =>
+      new AsyncSnapshot<T>(ConnectionState.none, null, null);
+
+  @override
+  AsyncSnapshot<T> onConnecting(AsyncSnapshot<T> current) =>
+      new AsyncSnapshot<T>(
+          ConnectionState.waiting, current.data, current.error);
+
+  @override
+  AsyncSnapshot<T> onData(AsyncSnapshot<T> current, T data) =>
+      new AsyncSnapshot<T>(ConnectionState.active, data, null);
+
+  @override
+  AsyncSnapshot<T> onError(AsyncSnapshot<T> current, dynamic error) =>
+      new AsyncSnapshot<T>(ConnectionState.active, null, error);
+
+  @override
+  AsyncSnapshot<T> onDone(AsyncSnapshot<T> current) =>
+      new AsyncSnapshot<T>(ConnectionState.done, current.data, current.error);
+
+  @override
+  AsyncSnapshot<T> onDisconnecting(AsyncSnapshot<T> current) =>
+      new AsyncSnapshot<T>(
+          ConnectionState.none, current.data, current.error);
+
+  @override
+  Widget build(BuildContext context, AsyncSnapshot<T> currentSummary) =>
+      builder(context, currentSummary);
+}
+
+/// Base class for Widgets that build themselves based on interaction with
+/// a specified [stream].
+///
+/// The widget's state maintains a summary of the interaction so far. The
+/// summary is defined by sub-classes.
+///
+/// [T] type of stream events.
+/// [S] type of summary.
+abstract class StreamFold<T, S> extends StatefulWidget {
+  final Stream<T> stream;
+
+  StreamFold({Key key, this.stream}) : super(key: key);
+
+  /// Returns the initial summary.
+  S initial();
+
+  /// Returns an updated version of the [current] summary reflecting that we
+  /// are now connected to a stream.
+  ///
+  /// The default implementation returns [current] as is.
+  S onConnecting(S current) => current;
```

Comment:
I am not sure I understand the exact semantics here but I'm guessing this should be didConnect, rather than onConnecting, based on the naming scheme described in the style guide?